refactor!: Surfaces: Change Transform3D storage by value

asalzburger opened the pull request:
This PR changes the memory model and consistently the construction of Surface/Layer/Volumes to by-value storage and construction from const reference.

Sits on top of #399 (merged)  and #405 (merged).

Fixes #408

asalzburger commented:
Comparison of the SurfaceIntersection bench mark:

master (w/o #399 and #405) :

```shell
Running 10 test cases...

Benchmarking theta=0.222428, phi=1.79499...
- Plane: 20000 runs of 2000 iteration(s), 275.6ms total, 13.4400+/-0.0006µs per run, 6.720+/-0.014ns per iteration
- Disc: 20000 runs of 2000 iteration(s), 364.7ms total, 17.2680+/-0.0011µs per run, 8.634+/-0.025ns per iteration
- Cylinder: 20000 runs of 2000 iteration(s), 1605.8ms total, 79.1810+/-0.0060µs per run, 39.590+/-0.133ns per iteration
- Straw: 20000 runs of 2000 iteration(s), 381.7ms total, 17.9860+/-0.0008µs per run, 8.993+/-0.017ns per iteration
```

With (#399, #405 and #411 ):

```shell 
Benchmarking theta=0.222428, phi=1.79499...
- Plane: 20000 runs of 2000 iteration(s), 268.7ms total, 13.2380+/-0.0005µs per run, 6.619+/-0.011ns per iteration
- Disc: 20000 runs of 2000 iteration(s), 334.2ms total, 16.5520+/-0.0004µs per run, 8.276+/-0.008ns per iteration
- Cylinder: 20000 runs of 2000 iteration(s), 1525.3ms total, 75.5540+/-0.0008µs per run, 37.777+/-0.018ns per iteration
- Straw: 20000 runs of 2000 iteration(s), 348.8ms total, 17.1430+/-0.0010µs per run, 8.572+/-0.022ns per iteration
```

Total time went from `105s` to `104s` - so overall looks good.

asalzburger commented:
Only the `BinUtility` remains for the moment with a `std::unique_ptr<const Transform3D>` for consistency will be likely changed as well in a subsequent PR.

paulgessinger commented:
@asalzburger this will require you to manually rebase/resolve conflicts.

asalzburger commented:
Rebased and waiting for a reviewer - @FabianKlimpel probably ?

FabianKlimpel commented:
> Rebased and waiting for a reviewer - @FabianKlimpel probably ?

I will have a look

FabianKlimpel commented on Tests/UnitTests/Core/Surfaces/ConeSurfaceTests.cpp:
```diff
@@ -34,17 +34,16 @@ BOOST_AUTO_TEST_SUITE(ConeSurfaces)
   // ConeSurface default constructor is deleted
   //
-  /// Constructor with transform pointer, null or valid, alpha and symmetry
+  /// Constructor with transform alpha and symmetry
```
```suggestion
  /// Constructor with transform, alpha and symmetry
```

asalzburger replied:
Fixed.

FabianKlimpel commented on Tests/UnitTests/Core/Surfaces/DiscSurfaceTests.cpp:
```diff
@@ -37,16 +37,13 @@ BOOST_AUTO_TEST_CASE(DiscSurfaceConstruction) {
   // scaffolding...
   double rMin(1.0), rMax(5.0), halfPhiSector(M_PI / 8.);
   //
-  /// Test DiscSurface fully specified constructor but no transform
-  BOOST_CHECK_NO_THROW(
-      Surface::makeShared<DiscSurface>(nullptr, rMin, rMax, halfPhiSector));
```
Why is this test removed?

asalzburger replied:
This was a special test for DiscSurface w/o transform, that doesn't exist anymore as no transform is not an option.

FabianKlimpel replied:
Alright

FabianKlimpel commented on Tests/UnitTests/Core/Surfaces/PerigeeSurfaceTests.cpp:
```diff
@@ -41,12 +41,9 @@ BOOST_AUTO_TEST_CASE(PerigeeSurfaceConstruction) {
-  BOOST_CHECK_EQUAL(Surface::makeShared<PerigeeSurface>(pNullTransform)->type(),
-                    Surface::Perigee);
+  auto pTransform = Transform3D(translation);
   BOOST_CHECK_EQUAL(Surface::makeShared<PerigeeSurface>(pTransform)->type(),
                     Surface::Perigee);
```
Here you changed the semantics of the test but I guess it's fine

asalzburger replied:
Because the old semantics doesn't exist anymore.

FabianKlimpel reviewed (changes requested):
I have a some (mostly small) comments/suggestions.
In general: these are very nice changes.

asalzburger commented on Core/include/Acts/Geometry/AbstractVolume.hpp:
```diff
@@ -44,18 +44,15 @@ class AbstractVolume : public Volume {
   /// Constructor with shared Transform3D*, VolumeBounds*
   ///
-  /// @param htrans is the transform 3D the positions the volume in global frame
+  /// @param transform is the transform 3D the positions the volume in global
+  /// frame
```
Fixed.

asalzburger commented on Core/include/Acts/Geometry/LayerCreator.hpp:
```diff
@@ -86,7 +80,7 @@ class LayerCreator {
       std::vector<std::shared_ptr<const Surface>> surfaces, size_t binsPhi,
       size_t binsZ, std::optional<ProtoLayer> _protoLayer = std::nullopt,
-      std::shared_ptr<const Transform3D> transform = nullptr,
+      const Transform3D& transform = s_idTransform,
```
I think we do not need s_idTransform anymore, as `Transform3D::Identity()` returns a static object as well, I make a PR to remove that, as `Transform3D::Identity()` is so much clearer.

FabianKlimpel replied:
I agree with you but at least it should be done consistently in the code base

asalzburger replied:
I opened #441

asalzburger commented on Core/include/Acts/Surfaces/CylinderSurface.hpp:
```diff
@@ -44,21 +44,22 @@ class CylinderSurface : public Surface {
-  /// @param htrans transform to position the surface, can be nullptr
-  /// @note if htrans == nullptr, the cylinder is positioned around (0.,0.,0.)
+  /// @param transform transform to position the surface
+  /// @note if transform == nullptr, the cylinder is positioned around
+  /// (0.,0.,0.)
```
Fixed.

asalzburger commented on Core/include/Acts/Surfaces/DiscSurface.hpp:
```diff
@@ -48,36 +48,33 @@ class DiscSurface : public Surface {
-  DiscSurface(std::shared_ptr<const Transform3D> htrans, double minhalfx,
-              double maxhalfx, double maxR, double minR, double avephi = 0.,
-              double stereo = 0.);
+  DiscSurface(const Transform3D& transform, double minhalfx, double maxhalfx,
+              double maxR, double minR, double avephi = 0., double stereo = 0.);
```
Why now, but this needs another PR, was it will touch the `TrapezoidDiscBounds` class.

FabianKlimpel replied:
I just saw it then and if it would have been a small change then it would have been fine to add it in this PR. If it's larger then for sure, another PR is needed.

asalzburger replied:
I opened an issue.

FabianKlimpel replied:
Thank you

asalzburger replied:
I opened #442 for this.

asalzburger commented on Core/src/Geometry/CylinderVolumeHelper.cpp:
```diff
@@ -51,19 +51,21 @@ Acts::CylinderVolumeHelper::createTrackingVolume(
-  // (3) !volumeBounds && transform  : estimate size from layers, use transform
-  // (4) !volumeBounds && !transform : estimate size & translation from layers
+  // (2) volumeBounds && transform==identity  : centered around 0, but with
+  // given bounds (3) !volumeBounds && transform  : estimate size from layers,
+  // use transform (4) !volumeBounds && transform==identity : estimate size &
```
Fixed.

asalzburger commented on Core/src/Geometry/LayerCreator.cpp:
```diff
@@ -148,9 +148,10 @@ Acts::MutableLayerPtr Acts::LayerCreator::cylinderLayer(
-    transform =
-        std::make_shared<const Transform3D>(Translation3D(0., 0., layerZ));
+  // create the layer transforms if not given
+  Translation3D addTranslation(0., 0., 0.);
+  if (transform.isApprox(s_idTransform) && bTypeZ == equidistant) {
+    addTranslation = Translation3D(0., 0., layerZ);
```
I think this one I will leave, it's more for clearer to read than the long tertiary operator.

asalzburger commented on Core/src/Geometry/SurfaceArrayCreator.cpp:
```diff
@@ -42,20 +42,18 @@ Acts::SurfaceArrayCreator::surfaceArrayOnCylinder(
-  Transform3D transform =
-      transformOpt != nullptr ? *transformOpt : Transform3D::Identity();
-
+  Transform3D ftransform = transform;
```
Because the transform is provided by reference to the parsing helper function that adapts it.

asalzburger commented on Core/src/Geometry/SurfaceArrayCreator.cpp:
```diff
@@ -42,20 +42,18 @@ Acts::SurfaceArrayCreator::surfaceArrayOnCylinder(
   // Transform lambda captures the transform matrix
-  auto globalToLocal = [transform](const Vector3D& pos) {
-    Vector3D loc = transform * pos;
+  auto globalToLocal = [ftransform](const Vector3D& pos) {
```
See above, the message signature of `create(...)` requires a non-const reference, hence the copy.

asalzburger commented on Core/include/Acts/Geometry/Volume.hpp:
```diff
@@ -33,26 +33,24 @@ class Volume : public virtual GeometryObject {
   ///
   /// @param vol is the source volume for the copy
-  /// @param shift is the optional shift applied after copying
+  /// @param shift is the optional shift applied as : shift * transform
```
Indeed, good suggestion, will do.

asalzburger commented on Core/include/Acts/Surfaces/ConeSurface.hpp:
```diff
@@ -67,9 +67,9 @@ class ConeSurface : public Surface {
   /// @param gctx The current geometry context object, e.g. alignment
   /// @param other is the source cone surface
-  /// @param transf is the additional transfrom applied after copying
+  /// @param shift is the additional transfrom applied after copying
```
I tried to do, it should always be shift.

asalzburger commented on Core/include/Acts/Surfaces/CylinderSurface.hpp:
```diff
@@ -70,9 +71,9 @@ class CylinderSurface : public Surface {
   /// @param gctx The current geometry context object, e.g. alignment
   /// @param other is the source cone surface
-  /// @param transf is the additional transfrom applied after copying
+  /// @param shift is the additional transfrom applied after copying
```
As said, should be `shift` everywhere

asalzburger commented on Core/src/Geometry/CylinderVolumeHelper.cpp:
```diff
@@ -800,12 +799,11 @@ void Acts::CylinderVolumeHelper::glueTrackingVolumes(
+    Transform3D transform = s_idTransform;
     if (std::abs(zMin + zMax) > 0.1) {
       // it's not a concentric cylinder, so create a transform
-      auto pTransform = std::make_shared<const Transform3D>(
-          Translation3D(Vector3D(0., 0., 0.5 * (zMin + zMax))));
-      transform = pTransform;
+      transform =
+          Transform3D(Translation3D(Vector3D(0., 0., 0.5 * (zMin + zMax))));
```
Yes, but it makes it difficult to read, I think I will leave it as is.

FabianKlimpel replied:
I hoped that clang format would make it easy to read^^ also, I guessed that this would reduce the amount of writings (at least a bit).

asalzburger commented on Core/src/Geometry/CylinderVolumeHelper.cpp:
```diff
@@ -936,10 +933,7 @@ Acts::CylinderVolumeHelper::createCylinderLayer(double z, double r,
                                                        << r);
   // positioning
-  std::shared_ptr<const Transform3D> transform =
-      (std::abs(z) > 0.1)
```
Because you need to create a `Transform3D` anyway, so it doesn't make sense to check to only create one when there's some random cut-off value.

FabianKlimpel replied:
Alright

asalzburger commented on Core/src/Geometry/CylinderVolumeHelper.cpp:
```diff
@@ -975,10 +969,7 @@ std::shared_ptr<const Acts::Layer> Acts::CylinderVolumeHelper::createDiscLayer(
 
   // positioning
-  std::shared_ptr<const Transform3D> transform =
-      (std::abs(z) > 0.1)
```
Same answer.

asalzburger commented on Core/src/Geometry/LayerCreator.cpp:
```diff
@@ -218,9 +219,9 @@ Acts::MutableLayerPtr Acts::LayerCreator::discLayer(
-  if (!transform) {
-    transform =
-        std::make_shared<const Transform3D>(Translation3D(0., 0., layerZ));
+  Translation3D addTranslation(0., 0., 0.);
+  if (transform.isApprox(s_idTransform)) {
+    addTranslation = Translation3D(0., 0., layerZ);
```
Same answer.

asalzburger commented on Core/src/Geometry/LayerCreator.cpp:
```diff
@@ -281,9 +282,9 @@ Acts::MutableLayerPtr Acts::LayerCreator::discLayer(
-  if (!transform) {
-    transform =
-        std::make_shared<const Transform3D>(Translation3D(0., 0., layerZ));
+  Translation3D addTranslation(0., 0., 0.);
+  if (transform.isApprox(s_idTransform)) {
+    addTranslation = Translation3D(0., 0., layerZ);
```
And again.

asalzburger commented on Core/src/Geometry/LayerCreator.cpp:
```diff
@@ -357,10 +359,10 @@ Acts::MutableLayerPtr Acts::LayerCreator::planeLayer(
-  if (!transform) {
+  Translation3D addTranslation(0., 0., 0.);
+  if (transform.isApprox(s_idTransform)) {
     // double shift = (layerZ + envZShift);
-    transform = std::make_shared<const Transform3D>(
-        Translation3D(centerX, centerY, centerZ));
+    addTranslation = Translation3D(centerX, centerY, centerZ);
```
Yup.

asalzburger commented on Core/src/Geometry/Volume.cpp:
```diff
@@ -20,48 +16,24 @@
       m_orientedBoundingBox(m_volumeBounds->boundingBox(
-          nullptr, {0.05_mm, 0.05_mm, 0.05_mm}, this)) {
-  if (htrans) {
-    m_center = htrans->translation();
```
Good call, I will have a look.

asalzburger replied:
Let's leave it for the moment, it would yet create another API change, as `center()` is returned by const ref.

asalzburger commented on Core/src/Geometry/Volume.cpp:
```diff
@@ -100,7 +69,7 @@ std::ostream& Acts::operator<<(std::ostream& sl, const Acts::Volume& vol) {
 Acts::Volume::BoundingBox Acts::Volume::boundingBox(
     const Vector3D& envelope) const {
-  return m_volumeBounds->boundingBox(m_transform.get(), envelope, this);
+  return m_volumeBounds->boundingBox(&m_transform, envelope, this);
```
Because I didn't want to load this PR with even more changes.

FabianKlimpel replied:
This is something I also just saw then and yes, keep it as it is

asalzburger commented on Core/src/Surfaces/PerigeeSurface.cpp:
```diff
@@ -13,22 +13,19 @@
-  Surface::m_transform = std::make_shared<const Transform3D>(
-      Translation3D(gp.x(), gp.y(), gp.z()));
-}
+    : LineSurface(Translation3D(gp.x(), gp.y(), gp.z()) * s_idTransform,
```
Changed.

asalzburger commented on Plugins/Digitization/src/CartesianSegmentation.cpp:
```diff
@@ -61,21 +61,19 @@ void Acts::CartesianSegmentation::createSegmentationSurfaces(
-      std::make_shared<Transform3D>(Transform3D::Identity());
-  auto mutableCounterPlaneTransform =
-      std::make_shared<Transform3D>(Transform3D::Identity());
+  auto readoutPlaneTransform = Transform3D::Identity();
+  auto counterPlaneTransform = Transform3D::Identity();
```
Yes, sure - it's just for code readability here.

asalzburger commented:
Hi @FabianKlimpel - I have rebased and addressed most of the comments, while I have moved a few others into new issues. Can you have a look if you are ok with these changes, then I can start working off the remaining points in new, smaller PRs.

asalzburger commented:
Overwriting the ongoing check of `build/measure`, as only cosmetic changes have been don.
